Iterative decoding with annotations

luanaruiz9 opened the pull request:
**Updated data_generation.py to output number of ops:** data_generation.py now produces an .ops file for the it_dec_test split containing the number of operations in each input sequence of the iterative decoding test split. The number of operations is used to inform the number of prediction loops necessary to decode the iterative decoding test samples.

**Updated train.py to use annotations in iterative decoding:** see config.use_annotations. Other changes include: including the possibility to add extra prediction loops (to the number of loops obtained from the data), see config.extra_loops; printing wrong predictions instead of random predictions; saving list of predictions in a pickle file; adding the possibility of using a saved model to only perform predictions, see config.just_do_pred.

**Updated default.py:** to include flags use_annotations (use annotated number of operations to inform the number of prediction loops at test time), extra_loops (number of loops in addition to those obtained from annotations) and just_do_pred (only do predictions from a saved model).

**Updated input_pipeline.py to handle annotations**

luanaruiz9 committed:
Updated data_generation.py to output number of ops

data_generation.py now produces an .ops file for the it_dec_test split containing the number of operations in each input sequence of the iterative decoding test split. The number of operations is used to inform the number of prediction loops necessary to decode the iterative decoding test samples.

luanaruiz9 committed:
Updated train.py to use annotations in iterative decoding

See config.use_annotations. Other changes include: including the possibility to add extra prediction loops (to the number of loops obtained from the data), see config.extra_loops; printing wrong predictions instead of random predictions; saving list of predictions in a pickle file; adding the possibility of using a saved model to only perform predictions, see config.just_do_pred.

luanaruiz9 committed:
Updated default.py

To include flags use_annotations (use annotated number of operations to inform the number of prediction loops at test time), extra_loops (number of loops in addition to those obtained from annotations) and just_do_pred (only do predictions from a saved model).

luanaruiz9 committed:
Updated input_pipeline.py to handle annotations

santiaontanon-google commented on datasets/pcfg/data_generation.py:
```diff
@@ -149,26 +148,31 @@ def main(argv: Sequence[str]):
     
     out_file_src = open(data_path + 'it_dec_val.src', 'w')
     out_file_tgt = open(data_path + 'it_dec_val.tgt', 'w')
```
Sorry for not catching this earlier. I Was struggling to understand the code below, since I did not understand what "out_file_src_dec" stood for. If I understand correctly "out_file_src" is for the "validation set" and "out_file_src_dec" for the test set, right? Maybe we should call them "out_file_src_validation" and "out_file_src_test" instead to make it more readable? (same with *_tgt)

santiaontanon-google commented on datasets/pcfg/data_generation.py:
```diff
@@ -149,26 +148,31 @@ def main(argv: Sequence[str]):
-    
     out_file_src_dec = open(data_path + 'it_dec_test.src', 'w')
     out_file_tgt_dec = open(data_path + 'it_dec_test.tgt', 'w')
+    ops_file = open(data_path + 'it_dec_test.ops', 'w')
```
Similarly, "ops_file" is only for test, so maybe call it "out_file_ops_test", to match with the other naming scheme?

santiaontanon-google commented on model/train.py:
```diff
@@ -588,99 +599,122 @@ def encode_tokens(labels):
-          checkpoints.save_checkpoint(workdir, jax_utils.unreplicate(optimizer),
-                                      step)
+  if config.just_do_pred:
+      exemplars, test_acc, predicted_list = decode_and_calculate_acc(
```
Maybe it's the Github visualization, but there some missing indentation here, no?

luanaruiz9 replied:
It looks like it was the Github visualization!

santiaontanon-google reviewed (approved):
Thanks Luana, just a few small nits

jtainslie commented on datasets/pcfg/data_generation.py:
```diff
@@ -118,10 +117,10 @@ def generate_intermediate_tasks(source_line, target_line):
   intermediate_tasks.append(last_task + ' END')
 
-  return intermediate_tasks
+  return intermediate_tasks, len(operations)
```
NIt: We should probably update the "Returns:" section of the docstring to mention `len(operations)`.

jtainslie commented on model/input_pipeline.py:
```diff
@@ -285,18 +285,20 @@ def filter_fn(x):
 
   if pack_examples:
-    dataset = pack_dataset(dataset, max_length)
+    dataset = pack_dataset(dataset, max_length, keys = ["inputs", "targets"])
```
Nit: single quotes to be consistent in this file.
```suggestion
    dataset = pack_dataset(dataset, max_length, keys = ['inputs', 'targets'])
```

jtainslie commented on model/train.py:
```diff
@@ -40,6 +40,7 @@
 import jax.numpy as jnp
 import ml_collections
 import numpy as np
+import pickle as pkl
```
Nit: This should appear with the standard library imports.  (https://google.github.io/styleguide/pyguide.html#313-imports-formatting)

jtainslie commented on model/train.py:
```diff
@@ -342,7 +343,9 @@ def tokens_ids_to_logits(flat_ids, flat_cache):
-  return np.concatenate([x, np.tile(x[-1], (batch_pad, 1))], axis=0)
+  if len(x.shape) == 1:
+      x = np.expand_dims(x, 1)
+  return np.concatenate([x, np.tile(x[-1], (batch_pad,1))], axis=0)
```
Formatting nit:
```suggestion
  return np.concatenate([x, np.tile(x[-1], (batch_pad, 1))], axis=0)
```

jtainslie commented on model/train.py:
```diff
@@ -415,6 +420,8 @@ def decode_and_calculate_acc(*, p_pred_step, p_init_cache, target, config,
     predicted = p_pred_step(pred_batch['inputs'], target, cache, decode.EOS_ID,
                             max_predict_length)
     predicted_list = [tohost(predicted)]
+    if use_annotations == True:
```
If `use_annotations` is expect to be boolean, then this would be simpler:
```suggestion
    if use_annotations:
```

jtainslie commented on model/train.py:
```diff
@@ -435,18 +442,22 @@ def decode_and_calculate_acc(*, p_pred_step, p_init_cache, target, config,
             predictions.append(decode_tokens(predicted[i]))
           if j == max_predict_loops - 1 and stop == False:
```
Nit: I should have caught this earlier, but it would be simpler to use:
```suggestion
          if j == max_predict_loops - 1 and not stop:
```
(Same with the other `stop == False` line above.)

luanaruiz9 committed:
Corrected nits in train.py

Plus, corrected predictions outputted by function decode_and_calculate_acc (now outputting all predictions --- over all prediction loops --- for all samples in all batches in translated form).

luanaruiz9 committed:
Corrected nit in input_pipeline.py

luanaruiz9 committed:
Corrected nits in data_generation.py
